### ISSUE TITLE
A proposal on catching weird non-popped errors

### ISSUE THREAD

**Issue by beldmit** [openssl/openssl#12897]
Trying to (unsuccessfully) fix the #12838 erroneous error stack, I got an idea.
There are also some places in the code where we refer to the emptiness of the error stack - e.g, returning the reason of SSL call failures.

What if we start to add an ID - just incremental uint32 - to any error message? 
If increment happens on adding error to stack it makes (possibly, per thread) more comfortable to catch the place where this particular error came from.

It can be done when a special debugging build option is specified.

**Comment by levitte** [comment]
> There are also some places in the code where we refer to the emptiness of the error stack

Really?  I haven't seen that...
That's pretty darn fragile, as that either trusts that one of our functions up the call stack has cleared the error log [^1], or that the application has.

> What if we start to add an ID - just incremental uint32 - to any error message?
> If increment happens on adding error to stack it makes (possibly, per thread) more comfortable to catch the place where this particular error came from.

We normally add debugging data: function name, filename, line in file
Isn't that enough?

However, there are a lot of places where we aren't adding an error record for the moment, and we should.  No matter of incremented identity will help that.  Could it be that you're hitting such a spot?

**Comment by beldmit** [comment]
> > There are also some places in the code where we refer to the emptiness of the error stack
> 
> Really? I haven't seen that...
> That's pretty darn fragile, as that either trusts that one of our functions up the call stack has cleared the error log [^1], or that the application has.

Sorry, I was wrong when speaking this.

I was speaking about https://github.com/openssl/openssl/blob/4bb73d5409c056a878f526280f86cc3c01f8cd68/ssl/ssl_lib.c#L3764-L3770

**Comment by beldmit** [comment]
> > What if we start to add an ID - just incremental uint32 - to any error message?
> > If increment happens on adding error to stack it makes (possibly, per thread) more comfortable to catch the place where this particular error came from.
> 
> We normally add debugging data: function name, filename, line in file
> Isn't that enough?

No. When I came across on non-empty error stack when we try to use the decoders sequentially, the information
```
00F19516BD7F0000:error::asn1 encoding routines:ASN1_get_object:header too long:crypto/asn1/asn1_lib.c:105:
```
is not helpful at all, because the line is triggered more than 1000 times

**Comment by nhorman** [comment]
@beldmit this seems untouched in quite some time, but the idea is interesting.  What I think you are proposing is an integer indicator that leaves a breadcrumb to what call path a given error was in when the error occured, right?  I'm not sure a single integer counter gets us there, but still, interesting

Do you have interest in pursuing this?

**Comment by nhorman** [comment]
no response, marking as inactive

**Comment by beldmit** [comment]
Sorry, missed your question. Yes, I think it's good enough idea, probably we can discuss it?

**Comment by nhorman** [comment]
so thinking about this, I think what your error indicator is meant to do is identify inconsistencies in the error log.  I.e. add a incremented counter to each error log entry such that the error entries have a visual marker when something didn't get previously cleared. i.e an error stack like:
```
ERROR A (id 10)
ERROR B (id 9)
ERROR C (id 1)
ERROR D (id 8)
```
would give you a hint that id 1 didn't get cleared from the stack prior to the call path that generated ids 10, 9 and 8.

that makes sense, but it seems like it could get confusing with adjacent calls (i.e. if you're interested in error id 11, but the prior call stack didn't get cleared, it may just look like one big call stack.

An improvement on that might be to augment all entry points to libcrypto/libssl with gcc's -finstrument-function switch, using the input from libcrypto.num (or perhaps strategically placed function attributes), such that, on those entry points a threads err counter is incremented.  That would result in something like:
```
ERR A(id 1)
ERR B(id 1)
ERR C(id 0)
ERR D(id 1)
```
Which would more clearly show the differentiation in call stacks.  That would be a fairly invasive change though, requiring lots of call site specific changes.  And it would be limited to gcc/clang

If we're ok with that limitation though, another option might be to create a conditional feature in debug builds that used the backtrace/backtrace_symbols call to record the call stack (or at least the last few entries) of every error.  We would need an additional api call to dump that info out, but it would make very clear the call path of every error, and could be implemented with the ERR_api entirely

**Comment by nhorman** [comment]
actually, as I think about this, that last suggestion might be better handled (and much more quickly handled) as a systemtap script.  You could add a probe in  to ERR_set_debug, and for every call made, record the data provided, along with the backtrace) to a buffer, to be printed on exit, which you could then filter at will

**Comment by nhorman** [comment]
So I was tinkering with this this weekend, and I really think systemtap is your answer here.

If you want to capture backtraces every time theres a error put on the stack, its pretty straightforward:
```
stap -d ./apps/openssl --ldd -e'\
probe process("./apps/openssl").library("./libcrypto.so.3").function("ERR_set_error") {printf("====================\n");\
print_ubacktrace() }' \
-c ./apps/openssl <subcommand>
```

Adjust as needed for your test case of course

**Comment by nhorman** [comment]
This isn't perfect, but it may help you out:
```
global err_id = 0;
global instance_counter = 0;
global backtraces[256];

probe process("/home/admin/git/openssl/apps/opnessl").library("/home/admin/git/openssl/libcrypto.so.3").function("ERR_set_error") {
        backtraces[instance_counter] = sprint_ubacktrace();
        instance_counter++;
}

probe process("/home/admin/git/openssl/apps/opnessl").library("/home/admin/git/openssl/libcrypto.so.3").function("ERR_clear_error") {
        printf("Clearning errors from stack\n");
        foreach (bt in backtraces) {
                printf("=================================\n");
                print_ustack(backtraces[bt]);
        }
        delete backtraces;
        instance_counter = 0;
        err_id++;
}
```

This script will show you the call stacks of every error that was raised between successive calls to ERR_clear_error.  Its a bit counter-intuitive, but it will let you see between successive clears if there are call stacks that don't belong to the same sets of operations.

**Comment by beldmit** [comment]
Thank you. I believe it's worth documenting this approach somewhere...

**Comment by nhorman** [comment]
I agree, though I'm not sure where that might fit well.  I'm working on a webinar that might discuss something like this, but I'm not sure thats a great canonical resource.  Do you have a suggestion for a location?

**Comment by beldmit** [comment]
Do we have Wiki alive?

**Comment by nhorman** [comment]
we do, but its currently in flux, though that would be a good place to document this when its final location is determined

**Comment by t8m** [comment]
I am closing this as this idea is unlikely to be ever implemented.

**Comment by beldmit** [comment]
It can be implemented but I'm currently not so eager doing this so agreed
